### ISSUE TITLE
[Bug] ONNX INT4 stops early on longer Chinese text; greedy decoding loops until max tokens

### ISSUE THREAD

**Issue by LVFENGCHUN** [Edge0-AI/Audio8_TTS#38]
# [Bug] ONNX INT4 stops early on longer Chinese text; ni'zagreedy decoding loops until max tokens

## Summary

`Audio8-TTS-Preview-0.6B-ONNX-INT4` consistently stops before reading all of a longer Chinese input with the default sampling settings. With an effectively greedy setting (`top_k=1`), it does not stop early, but repeats the beginning until `max_new_tokens` is reached.

The same reference voice and complete input work with the Transformers BF16 checkpoint. This may be related to the ONNX INT4 model, the ONNX sampling implementation, or EOS behavior after quantization.

## Environment

- OS: Linux x86_64
- CPU: Intel Xeon Platinum 8336C, 7 cores / 14 threads available
- Python: 3.12
- ONNX Runtime: 1.24.4
- Model: `Audio8/Audio8-TTS-Preview-0.6B-ONNX-INT4`
- Model fingerprint: `62dcff0adf6c2535b3260467a7c1d482b556da57266c96a444518b76e140d2c3`
- Precision: INT4 Slow/Fast AR, FP16 codec
- Threads: 5

## Reference voice

- File: `voice_04.wav` (available to attach)
- Format: 48 kHz, stereo, PCM 24-bit
- Duration: 2.346 seconds
- Exact transcript:

```text
刚才青叶公主在我们不方便动手
```

## Input text

```text
在人工智能快速发展的今天，语音合成技术正逐渐走进日常生活。它不仅可以帮助我们阅读新闻、播报通知，还能为有声内容和智能助手提供更加自然、清晰并且富有表现力的声音。现在，我们正在使用一段较长的中文文本，测试模型在连续表达、停顿节奏和音色保持方面的实际效果。
```

## Steps to reproduce

Download the model and set up the official ONNX runtime as documented:

```bash
cd onnx_runtime
hf download Audio8/Audio8-TTS-Preview-0.6B-ONNX-INT4 --local-dir model
bash setup.sh
bash start_server.sh
```

Register the reference voice:

```bash
curl http://127.0.0.1:8024/api/voices/register \
  -F 'audio=@/absolute/path/voice_04.wav' \
  -F 'text=刚才青叶公主在我们不方便动手' \
  -F 'name=voice_04' \
  -F 'overwrite=true'
```

Run inference with the default sampling settings:

```bash
bash run_infer.sh \
  --voice voice_04 \
  --max-new-tokens 1024 \
  --text '在人工智能快速发展的今天，语音合成技术正逐渐走进日常生活。它不仅可以帮助我们阅读新闻、播报通知，还能为有声内容和智能助手提供更加自然、清晰并且富有表现力的声音。现在，我们正在使用一段较长的中文文本，测试模型在连续表达、停顿节奏和音色保持方面的实际效果。' \
  --output outputs/long_default.wav
```

The default CLI settings are `temperature=0.7`, `top_p=0.9`, `top_k=50`, and `seed=42`.

For the greedy-like test, run the same request with `--top-k 1`.

### Default sampling on CPU

- Generated frames: 130
- Audio duration: 6.037 seconds
- The output stops after the first sentence.
- ASR transcription of the output:

```text
在人工智能快速发展的今天
语音合成技术正逐渐走进日常生活
```

This result was identical across three runs with the fixed seed.

### `top_k=1` on CPU

- Generated frames: 1024 (hits `max_new_tokens`)
- Audio duration: 47.554 seconds
- The output repeatedly speaks the beginning instead of finishing the input.
- The result was identical across three runs.

Approximate ASR transcription:

```text
在人工智能快速发展的今天
我们正在使用一段更长的中文文本来测试语音合成模型……
在人工智能快速发展的今天……
```

### Additional CUDA Execution Provider observation

I also tested the same ONNX model experimentally with ONNX Runtime GPU 1.24.4. This is not part of the official CPU-only runtime and required changing the session provider.

With the default sampling settings, it generated 367 frames / 17.043 seconds and stopped after the second sentence. This suggests that small numerical differences between execution providers change where EOS is selected, but the full input is still not completed.

### Transformers BF16 comparison

The Transformers BF16 checkpoint generated the complete input with the same reference voice. It produced about 23.59 seconds of audio, without the early truncation seen in ONNX INT4.

I also ran a reference-free control with the Transformers BF16 checkpoint. It completed the same three-sentence input, produced 513 frames / 23.824 seconds of audio, and ASR confirmed that the final sentence was present. Two warm runs were byte-identical.

The official ONNX runtime does not currently permit the equivalent reference-free control: the CLI requires `--voice`, `VoiceStore` rejects an empty reference, and `PromptBuilder` requires reference codes with shape `[10, T>0]`. Therefore I could isolate the failure to the ONNX path versus the base BF16 model, but could not determine whether ONNX reference conditioning specifically contributes to the issue.

## Expected behavior

The model should read the complete input and then emit EOS. Changing from sampled decoding to greedy-like decoding should not turn early termination into an infinite/repetitive sequence that only stops at `max_new_tokens`.

## Questions

1. Is long-text early EOS a known limitation of the ONNX INT4 checkpoint?
2. Are there recommended sampling settings for reliable Chinese long-form synthesis?
3. Could quantization be changing the EOS logits enough to cause this behavior?
4. Would it be possible for the runtime to report whether generation ended via EOS or `max_new_tokens`?

[voice_04.wav](https://github.com/user-attachments/files/31777008/voice_04.wav)
[run_2_onnx_greedy.wav](https://github.com/user-attachments/files/31777049/run_2_onnx_greedy.wav)
[run_2_onnx.wav](https://github.com/user-attachments/files/31777050/run_2_onnx.wav)

**Comment by Alichua** [comment]
Hi @LVFENGCHUN,

Thanks for the detailed report and reproducible test case. We confirmed this behavior locally.

This is not an intended long-text limitation. It appears to be a failure mode of the INT4 ONNX path: small numerical/logit changes can make EOS win too early, while top_k=1 removes sampling diversity and can lead to a repeated-token loop.

As a temporary workaround,  you may try this config:
temperature=0.3
top_p=0.9
top_k=10
(top_k=20 is also worth trying if top_k=10 behaves too deterministic) 

Lower sampling variance makes the semantic and codec token sequence more stable in our tests. For very long text, splitting it into shorter utterances also helps. On the other hand, top_k=1/greedy mode should be avoid, which may loop until max_new_tokens under the INT4 version.

Quantization and small execution-provider differences can change the EOS logit ranking. Our CPU/GPU observations are consistent with this, although we cannot yet completely separate quantization effects from other ONNX runtime differences. 

We also noticed that some long outputs gradually lose volume toward the end. The settings above can reduce this unexpected behavior, too.

Thanks again for the detailed report and test materials!
